Buildings/Fluid/CHPs/BaseClasses/EnergyConversion.mo

```diff
@@ -0,0 +1,196 @@
+      iconTransformation(extent={{100,-50},{140,-10}})));
+  Buildings.Controls.OBC.CDL.Interfaces.RealOutput QGen(
+    final unit="W") "Heat generation"
+    annotation (Placement(transformation(extent={{140,-120},{180,-80}}),
+      iconTransformation(extent={{100,-100},{140,-60}})));
+
+protected
+  Buildings.Fluid.CHPs.BaseClasses.AssertFuel assFue(final per=per)
+    "Assert if fuel flow rate is outside boundaries"
+    annotation (Placement(transformation(extent={{120,40},{140,60}})));
+  Buildings.Fluid.CHPs.BaseClasses.OperModeBasic opeModBas(final per=per)
+    "Typical energy conversion mode"
+    annotation (Placement(transformation(extent={{-20,-20},{0,0}})));
+  Buildings.Fluid.CHPs.BaseClasses.OperModeWarmUpEngTem opeModWarUpEngTem(
+    final per=per) if per.warmUpByTimeDelay
+    "Warm-up by engine temperature"
+    annotation (Placement(transformation(extent={{-20,-60},{0,-40}})));
+  Buildings.Controls.OBC.CDL.Continuous.Sources.Constant const(final k=0)
+    "Zero constant"
+    annotation (Placement(transformation(extent={{-120,-10},{-100,10}})));
+  Buildings.Controls.OBC.CDL.Logical.Switch switch
+    annotation (Placement(transformation(extent={{-60,10},{-40,30}})));
+  Modelica.Blocks.Sources.BooleanExpression booExp(
+    final y=opeMod ==CHPs.BaseClasses.Types.Mode.WarmUp or
+            opeMod ==CHPs.BaseClasses.Types.Mode.Normal)
+    "Check if warm up mode or normal operation mode"
+    annotation (Placement(transformation(extent={{-120,10},{-100,30}})));
+  Modelica.Blocks.Sources.BooleanExpression booExp1(
+    final y=opeMod ==CHPs.BaseClasses.Types.Mode.WarmUp and
+            per.warmUpByTimeDelay) "Check if warm up mode by time delay"
+    annotation (Placement(transformation(extent={{-60,40},{-40,60}})));
+  Buildings.Controls.OBC.CDL.Continuous.Sources.Constant const1(final k=0)
+    "Zero constant"
+    annotation (Placement(transformation(extent={{-80,70},{-60,90}})));
+  Buildings.Controls.OBC.CDL.Logical.Switch switch1
+    annotation (Placement(transformation(extent={{-20,40},{0,60}})));
+  Modelica.Blocks.Sources.BooleanExpression booExp3(
+    final y=opeMod ==CHPs.BaseClasses.Types.Mode.WarmUp and
+          not per.warmUpByTimeDelay)
+    "Check if it is typical operation or warm-up by engine temperature"
+    annotation (Placement(transformation(extent={{40,40},{60,60}})));
+  Buildings.Controls.OBC.CDL.Logical.Switch switch2
+    annotation (Placement(transformation(extent={{80,70},{100,90}})));
+  Buildings.Controls.OBC.CDL.Logical.Switch switch3
+    annotation (Placement(transformation(extent={{80,10},{100,30}})));
+  Buildings.Controls.OBC.CDL.Logical.Switch switch4
+    annotation (Placement(transformation(extent={{80,-50},{100,-30}})));
+  Buildings.Controls.OBC.CDL.Logical.Switch switch5
+    annotation (Placement(transformation(extent={{80,-110},{100,-90}})));
+  Buildings.Controls.OBC.CDL.Continuous.Sources.Constant Troo(
```

Comment:
`TRoo` is a constant here whereas it is provided through the heat port in the top level model for the computation of the engine temperature.